File: Add Code Here/Dijkstra's algorithm.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <limits>

using namespace std;

const int INF = numeric_limits<int>::max();

class Graph {
public:
    int V; // Number of vertices
    vector<vector<pair<int, int>> > adj; // Adjacency list

    Graph(int V) {
        this->V = V;
        adj.resize(V);
    }

    void addEdge(int u, int v, int w) {
        adj[u].push_back({v, w});
        adj[v].push_back({u, w}); // For undirected graphs
    }

    void dijkstra(int src) {
        vector<int> distance(V, INF);
        distance[src] = 0;

        // Create a min-heap to store vertices with their distances
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
        minHeap.push({0, src});

        while (!minHeap.empty()) {
            int u = minHeap.top().second;
            minHeap.pop();

            for (const auto& neighbor : adj[u]) {
                int v = neighbor.first;
                int weight = neighbor.second;

                // Relaxation step
                if (distance[v] > distance[u] + weight) {
                    distance[v] = distance[u] + weight;
                    minHeap.push({distance[v], v});
                }
            }
        }

        // Print the shortest distances from the source vertex
        cout << "Shortest distances from source " << src << " to all other vertices:" << endl;
        for (int i = 0; i < V; i++) {
            cout << "Vertex " << i << ": " << distance[i] << endl;
        }
    }
};

int main() {
    int V = 6; // Number of vertices
    Graph g(V);

    // Adding edges with weights
    g.addEdge(0, 1, 2);
    g.addEdge(0, 2, 4);
    g.addEdge(1, 2, 1);
    g.addEdge(1, 3, 7);
    g.addEdge(2, 4, 3);
    g.addEdge(3, 4, 2);
    g.addEdge(3, 5, 1);
    g.addEdge(4, 5, 5);

    int sourceVertex = 0; // Source vertex for Dijkstra's algorithm
    g.dijkstra(sourceVertex);

    return 0;
}

File: Add Code Here/PYTHON/Dijkstra'sAlgorithm.py
import heapq

class Graph:
    def __init__(self):
        self.vertices = set()
        self.edges = {}
    
    def add_edge(self, start, end, weight):
        self.vertices.add(start)
        self.vertices.add(end)
        if start in self.edges:
            self.edges[start].append((end, weight))
        else:
            self.edges[start] = [(end, weight)]

def dijkstra(graph, start):
    shortest_path = {vertex: float('infinity') for vertex in graph.vertices}
    shortest_path[start] = 0
    priority_queue = [(0, start)]

    while priority_queue:
        current_dist, current_vertex = heapq.heappop(priority_queue)

        if current_dist > shortest_path[current_vertex]:
            continue

        if current_vertex in graph.edges:
            for neighbor, weight in graph.edges[current_vertex]:
                distance = current_dist + weight
                if distance < shortest_path[neighbor]:
                    shortest_path[neighbor] = distance
                    heapq.heappush(priority_queue, (distance, neighbor))

    return shortest_path

def find_shortest_path(graph, start, end):
    distances = dijkstra(graph, start)
    path = [end]
    while end != start:
        for neighbor, _ in graph.edges[end]:
            if distances[neighbor] + graph.edges[neighbor][0][1] == distances[end]:
                path.append(neighbor)
                end = neighbor
                break
    return path[::-1]

if __name__ == "__main__":
    graph = Graph()
    graph.add_edge("A", "B", 1)
    graph.add_edge("A", "C", 4)
    graph.add_edge("B", "C", 2)
    graph.add_edge("B", "D", 5)
    graph.add_edge("C", "D", 1)
    graph.add_edge("D", "E", 3)

    start_vertex = "A"
    end_vertex = "E"

    shortest_path = find_shortest_path(graph, start_vertex, end_vertex)
    print("Shortest Path:", shortest_path)

